numba/tests/test_typedlist.py

```diff
@@ -155,6 +155,30 @@ def test_unsigned_access(self):
         self.assertEqual(L.pop(ui32_1), 2)
         self.assertEqual(L.pop(ui32_0), 123)
 
+    def test_dtype(self):
```

Comment:
You can probably slim this test down a bit, I would suggest something along the lines of the following:

```python
@njit 
def foo():
    li, lf = List(), List()
    li.append(int32(1))
    lf.append(float32(1.0))
    return li.dtype, lf.dtype

for func in foo, foo.py_func:
    self.assertEqual(func(), (int32, float32))
```

This way you write the test for int + float plus njit and python function in one go. Also you avoid any issue with testing across both32 and 64 bit archs.

Reply:
Thank you for the comments, I have modified and update this PR. 
However, I just realised that the `dtype` returned by `foo` is of the class `numpy.dtype`.
```python
>>> type(foo())
<class 'numpy.dtype'>
```
Also, I think the following result is a little non-intuitive.
```python
@njit
def foo():
    x = List()
    x.append(float32(3.3))
    return x.dtype

assert foo() == float32
```
The above assertion throws an error
```
Traceback (most recent call last):
  File "test_dtype.py", line 14, in <module>
    assert foo() == float32
TypeError: data type not understood
```
I want to know if this is the expected behaviour?

Reply:
I don't think that error should be raised, and I don't think that `foo` should be returning NumPy dtypes.

Reply:
The 'casting' of Numba dtypes to NumPy dtypes seems unrelated to the typed list:

```
In [1]: from numba import njit, int32

In [2]: @njit
   ...: def foo():
   ...:     return int32
   ...:

In [3]: foo()
Out[3]: dtype('int32')

In [4]: a = foo()

In [5]: type(a)
Out[5]: numpy.dtype
```